src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -503,5 +503,92 @@ protected override void OnMarked(NodeFactory context)
         {
             //Debug.Assert(_type.IsTypeDefinition || !_type.HasSameTypeDefinition(context.ArrayOfTClass), "Asking for Array<T> EEType");
         }
+
+        /// <summary>
+        /// Validates that it will be possible to create an EEType for '<paramref name="type"/>'.
+        /// </summary>
+        public static void CheckCanGenerateEEType(TypeDesc type)
+        {
+            // TODO: a set of rules for "constructed" types: e.g. ban creating a constructed EEType for '<Module>' type,
+            //       pointer types, function pointer types, byrefs, generic definitions, etc.
+
+            // It must be possible to create an EEType for the base type of this type
+            TypeDesc baseType = type.BaseType;
+            if (baseType != null)
+            {
+                CheckCanGenerateEEType(baseType);
+            }
+
+            // Don't validate generic definitons
+            if (type.IsGenericDefinition)
+            {
+                return;
+            }
+
+            // We need EETypes for interfaces
+            foreach (var intf in type.RuntimeInterfaces)
+            {
+                CheckCanGenerateEEType(intf);
+            }
+
+            // Validate classes, structs, enums, interfaces, and delegates
+            DefType defType = type as DefType;
+            if (defType != null)
+            {
+                // Ensure we can compute the type layout
+                defType.ComputeInstanceLayout(InstanceLayoutKind.TypeAndFields);
+                defType.ComputeStaticFieldLayout(StaticLayoutKind.StaticRegionSizesAndFields);
+
+                // Make sure instantiation length matches the expectation
+                // TODO: it might be more resonable for the type system to enforce this (also for methods)
+                if (defType.Instantiation.Length != defType.GetTypeDefinition().Instantiation.Length)
+                {
+                    throw new TypeSystemException.TypeLoadException(ExceptionStringID.ClassLoadGeneral, type);
+                }
+
+                foreach (TypeDesc typeArg in defType.Instantiation)
+                {
+                    // ByRefs, pointers, function pointers, and System.Void are never valid instantiation arguments
+                    if (typeArg.IsByRef || typeArg.IsPointer || typeArg.IsFunctionPointer || typeArg.IsVoid)
```

Comment:
We should be assuming that we are given valid IL. This kind of verification may be in asserts in the type system; or in the reflection argument validation; or in the PEVerify-like tool.

The JIT does not do any verification of the IL either. If you feed it invalid IL, it will generate bad code, or just plain crash. We should be holding the type system to the same bar.